src/tools/llvm-bitcode-linker/src/linker.rs

```diff
@@ -0,0 +1,240 @@
+use std::path::Path;
+use std::path::PathBuf;
+
+use anyhow::Context;
+
+use crate::Optimization;
+use crate::Target;
+
+#[derive(Debug)]
+pub struct Session {
+    target: Target,
+    cpu: Option<String>,
+    symbols: Vec<String>,
+    bitcode: Vec<PathBuf>,
+
+    // Output files
+    link_path: PathBuf,
+    opt_path: PathBuf,
+    sym_path: PathBuf,
+    out_path: PathBuf,
+}
+
+impl Session {
+    pub fn new(target: crate::Target, cpu: Option<String>, out_path: PathBuf) -> Self {
+        let link_path = out_path.with_extension("o");
+        let opt_path = out_path.with_extension("optimized.o");
+        let sym_path = out_path.with_extension("symbols.txt");
+
+        Session {
+            target,
+            cpu,
+            symbols: Vec::new(),
+            bitcode: Vec::new(),
+            link_path,
+            opt_path,
+            sym_path,
+            out_path,
+        }
+    }
+
+    /// Link a rlib into a bitcode object and add it to the list of files ready to be linked
+    pub fn link_rlib(&mut self, path: impl AsRef<Path>, keep_symbols: bool) -> anyhow::Result<()> {
+        let output_file_link = path.as_ref().with_extension("o");
+        tracing::info!(
+            "Linking rlib: {} into bitcode: {}",
+            path.as_ref().display(),
+            output_file_link.display(),
+        );
+
+        let link_output = std::process::Command::new("llvm-link")
+            .arg(path.as_ref())
+            .arg("-o")
+            .arg(&output_file_link)
+            .arg("--ignore-non-bitcode")
+            .output()
+            .unwrap();
+
+        if !link_output.status.success() {
+            tracing::error!(
+                "llvm-link returned with Exit status: {}\n stdout: {}\n stderr: {}",
+                link_output.status,
+                String::from_utf8(link_output.stdout).unwrap(),
+                String::from_utf8(link_output.stderr).unwrap(),
+            );
+            anyhow::bail!("llvm-link failed to link file {}", path.as_ref().display());
+        }
+
+        self.add_bitcode(output_file_link, keep_symbols)
+    }
+
+    /// Add a single symbol to the list of exported symbols
+    pub fn add_exported_symbol(&mut self, symbol: String) {
+        self.symbols.push(symbol);
+    }
+
+    /// Add a Vec of symbols to the list of exported symbols
+    pub fn add_exported_symbols(&mut self, symbols: Vec<String>) {
+        self.symbols.extend(symbols);
+    }
+
+    /// Add a bitcode module ready to be linked
+    pub fn add_bitcode(
+        &mut self,
+        path: impl AsRef<Path>,
+        keep_symbols: bool,
```

Comment:
Done. I kept the function to avoid having to expose the `Vec<_>` of filepaths directly. On the other side I merged it with the rlib method so one method was removed. Hope this was a fine solution.